src/parser.rs

```diff
@@ -59,106 +104,528 @@ impl Header {
+            let entry_type = reader.read_u8()?;
+            let index1 = reader.read_u8()?;
+            let index2 = reader.read_u8()?;
+            let num = reader.read_u8()?;
+            let num1 = (num >> 4) & 0x0F;
+            let num2 = num & 0x0F;
+            let service_id = reader.read_u16::<BE>()?;
+            let instance_id = reader.read_u16::<BE>()?;
+            let major_version = reader.read_u8()?;
+            let ttl = u24::new(reader.read_u24::<BE>()?);
+
+            let options = SdOptionRef {
+                index1,
+                index2,
+                num1,
+                num2,
+            };
+
+            entries.push(match SdEntry::is_service_entry(entry_type) {
+                true => {
+                    let minor_version = reader.read_u32::<BE>()?;
+                    SdEntry::from_service(
+                        entry_type,
+                        SdServiceEntry {
+                            service_id,
+                            instance_id,
+                            major_version,
+                            minor_version,
+                            ttl,
+                            options,
+                        },
+                    )?
+                }
+                false => {
+                    reader.read_u16::<BE>()?; // reserved
+                    let eventgroup_id = reader.read_u16::<BE>()?;
+                    SdEntry::from_eventgroup(
+                        entry_type,
+                        SdEventgroupEntry {
+                            service_id,
+                            eventgroup_id,
+                            instance_id,
+                            major_version,
+                            ttl,
+                            options,
+                        },
+                    )?
                 }
+            });
+
+            read += SD_ENTRY_LENGTH;
+        }
+
+        let mut options: Vec<SdOption> = Vec::new();
+        let mut read: usize = 0;
+        let length: usize = match reader.read_u32::<BE>() {
+            Ok(value) => value as usize,
+            Err(_) => 0usize,
+        };
+
+        while read < length {
+            let option_len: usize = reader.read_u16::<BE>()? as usize;
+            let option_type = reader.read_u8()?;
+
+            if !SdOption::is_ip_option(option_type) {
+                for _ in 0..option_len {
+                    reader.read_u8()?; // drop option
+                }
+                read = read + 3 + option_len;
+                continue;
             }
-            Err(e) => Err(e),
+
+            reader.read_u8()?; // reserved
+
+            let ip: IpAddr = match SdOption::is_ip4_option(option_type) {
+                true => IpAddr::V4(Ipv4Addr::new(
+                    reader.read_u8()?,
+                    reader.read_u8()?,
+                    reader.read_u8()?,
+                    reader.read_u8()?,
+                )),
+                false => IpAddr::V6(Ipv6Addr::new(
```

Comment:
https://doc.rust-lang.org/std/net/struct.Ipv6Addr.html#impl-From%3C%5Bu16%3B%208%5D%3E is faster